### PR TITLE
Initial Chaos Test integration for SharedInformers

### PR DESCRIPTION
**Pull request by andreaTP**
## Description

As promised in those tests we are running Chaos Mesh to verify that the `SharedInformer` successfully recovers after various network failures.

We did the negative test and those tests would have caught early issues like #5189

## Type of change
 - [ ] Bug fix (non-breaking change which fixes an issue)
 - [ ] Feature (non-breaking change which adds functionality)
 - [ ] Breaking change (fix or feature that would cause existing functionality to change
 - [X] Chore (non-breaking change which doesn't affect codebase;
   test, version modification, documentation, etc.)

## Checklist
 - [X] Code contributed by me aligns with current project license: [Apache 2.0](https://www.apache.org/licenses/LICENSE-2.0)
 - [ ] I Added [CHANGELOG](https://github.com/fabric8io/kubernetes-client/blob/master/CHANGELOG.md) entry regarding this change
 - [X] I have implemented unit tests to cover my changes
 - [ ] I have added/updated the [javadocs](https://www.javadoc.io/doc/io.fabric8/kubernetes-client/latest/index.html) and other [documentation](https://github.com/fabric8io/kubernetes-client/blob/master/doc/CHEATSHEET.md) accordingly
 - [ ] No new bugs, code smells, etc. in [SonarCloud](https://sonarcloud.io/dashboard?id=fabric8io_kubernetes-client) report
 - [X] I tested my code in Kubernetes
 - [ ] I tested my code in OpenShift

### CODE REVIEW

File: chaos-tests/src/test/java/ChaosIT.java [hunk 1]
```diff
@@ -0,0 +1,161 @@
...
+      System.out.println("*** Checker Logs ***\n" + checkerLogs);
+      System.out.println("******");
+    } catch (Exception e) {
+      // ignore
+    }
+
+    try {
+      String controlLogs = controlSelector().getLog();
+      System.out.println("*** Control Logs ***\n" + controlLogs);
+      System.out.println("******");
```
Comment (rohanKanojia) [inline chaos-tests/src/test/java/ChaosIT.java:75]: In some places I see Slf4j logger getting used, in some places `System.out` . Is it possible to use only one way?

File: chaos-tests/pom.xml [hunk 2]
```diff
@@ -0,0 +1,228 @@
...
+    <dependency>
+      <groupId>org.junit.jupiter</groupId>
+      <artifactId>junit-jupiter-engine</artifactId>
+      <scope>test</scope>
+    </dependency>
+
+    <dependency>
+      <groupId>org.assertj</groupId>
+      <artifactId>assertj-core</artifactId>
+    </dependency>
```
Comment (rohanKanojia) [inline chaos-tests/pom.xml:88]: This dependency doesn't seem to be getting used

File: chaos-tests/src/main/java/io/fabric8/it/CheckerCommand.java [hunk 3]
```diff
@@ -0,0 +1,142 @@
...
+          throw new RuntimeException("Expected " + num + " but reached " + counter.get());
+        }
+      }
+    };
+
+    Map<String, String> labels = new HashMap<>();
+    labels.put(labelKey, labelValue);
+    Map<String, String> data = new HashMap<>();
+    data.put("counter", Integer.toString(0));
+    ConfigMap defaultConfigMap = new ConfigMapBuilder()
```
Comment (rohanKanojia) [inline chaos-tests/src/main/java/io/fabric8/it/CheckerCommand.java:115]: nit, unused variable

File: chaos-tests/pom.xml [hunk 4]
```diff
@@ -0,0 +1,228 @@
...
+        <artifactId>kubernetes-maven-plugin</artifactId>
+        <version>${jkube.version}</version>
+        <configuration>
+          <images>
+            <image>
+              <name>chaos-test-${app-name}:latest</name>
+              <build>
+                <from>openjdk:latest</from>
+                <cmd>java -jar /maven/${project.artifactId}-${project.version}.jar</cmd>
+              </build>
```
[the remark this answers is not in the document]
Comment (andreaTP) [reply chaos-tests/pom.xml:153]: > Is it to override image name?

yes, that's one reason.

> Why have you defined your own image configuration?

First time using `jkube` here, I started from a hello-world and that's the configuration that worked, happy to hear suggestions!

File: chaos-tests/pom.xml (hunk 2 above)
Comment (andreaTP) [reply chaos-tests/pom.xml:88]: 👍

File: chaos-tests/src/main/java/io/fabric8/it/CheckerCommand.java (hunk 3 above)
Comment (andreaTP) [reply chaos-tests/src/main/java/io/fabric8/it/CheckerCommand.java:115]: 👍

File: chaos-tests/pom.xml (hunk 4 above)
[the remark this answers is not in the document]
Comment (rohanKanojia) [reply chaos-tests/pom.xml:148]: Could you please try out these properties:
-  `jkube.generator.name` property to override default image name
-  `jkube.generator.from` to override base image.

If you get error for resolving mainClass, you can try providing `jkube.generator.java-exec.mainClass` in checker and control profiles.

[the remark this answers is not in the document]
Comment (andreaTP) [reply chaos-tests/pom.xml:148]: Hey @rohanKanojia ! I have attempted a few permutations:
- removing the `<build>` block makes the mangling of the image name ineffective but the resulting image works
- adding an empty `<build>` block the name gets mangled accordingly but the image doesn't work (java is not installed?)

All in all, I do consider those changes nitpicks, as we are discussing the permutation of 2 lines of XML 🙂 .
I propose to file an issue in the jkube repo, with a minimal repro and move there this discussion, wdyt?

[the remark this answers is not in the document]
Comment (rohanKanojia) [reply chaos-tests/pom.xml:148]: Sorry for not being clear. I meant removing the whole plugin configuration altogether and trying properties.

JKube considers an `<image>` configuration without `<build>` as invalid

> All in all, I do consider those changes nitpicks, as we are discussing the permutation of 2 lines of XML slightly_smiling_face 

You're right. Feel free to mark this conversation as resolved.

> I propose to file an issue in the jkube repo, with a minimal repro and move there this discussion, wdyt?

:+1:

**Comment by sunix** [comment]
Hey, are these tests been triggered somewhere in a github action ?

**Comment by andreaTP** [comment]
@sunix only after we merge they are going to be triggered.

**Comment by sunix** [comment]
> @sunix only after we merge they are going to be triggered.

Yeah, I was just wondering if you had triggered it in your personal repo that we could see

**Comment by andreaTP** [comment]
> Yeah, I was just wondering if you had triggered it in your personal repo that we could see

My runs has been too old now, re-triggered:
https://github.com/andreaTP/kubernetes-client/actions/runs/5575025561

File: chaos-tests/src/test/resources/checker-pod.yaml [hunk 5]
```diff
@@ -0,0 +1,29 @@
...
+apiVersion: v1
+kind: Pod
+metadata:
+  name: checker
+  labels:
+    app: checker
+spec:
+  containers:
+  - name: checker
+    image: docker.io/andreatp/fabric8-it-checker-app:latest
```
Comment (sunix) [inline chaos-tests/src/test/resources/checker-pod.yaml:26]: will that work ?

Comment (sunix) [reply chaos-tests/src/test/resources/checker-pod.yaml:26]: ok it is replaced in ChaosIT

File: chaos-tests/Readme.md [hunk 6]
```diff
@@ -0,0 +1,42 @@
+# Chaos Mesh tests for Kubernetes Client SharedInformer
+
+This module will run automated Chaos Tests for the SharedInformers
```
Comment (sunix) [inline chaos-tests/Readme.md:3]: It would be also nice to have a summary here of what the test is doing:
- what in this ChaosMesh is doing ?
- high level description of what and how the Checker and Control are working (explain the counter story?)

File: chaos-tests/src/test/resources/checker-pod.yaml (hunk 5 above)
Comment (andreaTP) [reply chaos-tests/src/test/resources/checker-pod.yaml:26]: yup, this is just a placeholder ...

File: chaos-tests/Readme.md (hunk 6 above)
Comment (sunix) [reply chaos-tests/Readme.md:3]: Thanks this is just perfect, I think it is worth adding this in the Readme or in ChaosIT javadoc. I agree this is kind of a translation in English but at that stage, the whole logic is located in several places which makes the story a bit hard to follow.

Comment (andreaTP) [reply chaos-tests/Readme.md:3]: @sunix added something to the Readme, let me know if you think it's enough.

Comment (sunix) [reply chaos-tests/Readme.md:3]: Have you push your changes?

Comment (andreaTP) [reply chaos-tests/Readme.md:3]: ooops, wrong branch 🤦 done now!

**Review by sunix — APPROVED** [review]

File: chaos-tests/Readme.md (hunk 6 above)
Comment (sunix) [reply chaos-tests/Readme.md:3]: Perfect thanks !!!

**Comment by manusa** [comment]
I'll try to check/review the PR today.

**Comment by andreaTP** [comment]
@manusa have you had a chance to review? Just rebased on latest `main`

**Comment by manusa** [comment]
> @manusa have you had a chance to review? Just rebased on latest `main`

Next in m yqueue

**Review by manusa — APPROVED** [review]
